TU20Bot/Configuration/AuthorizationModule.cs

```diff
@@ -0,0 +1,155 @@
+
+                if (string.IsNullOrEmpty(result)) {
+                    error("ERROR: Empty payload.", context.Response);
+                    return;
+                }
+
+                // Deserialize the JWT Token into an AuthorizationPayload object
+                var payload = JsonConvert.DeserializeObject<AuthorizationPayload>(result);
+
+                if (payload == null) {
+                    error("ERROR: Invalid payload.", context.Response);
+                    return;
+                }
+
+                // If the token is expired
+                if (DateTime.Now.CompareTo(payload.validUntil) > 0) {
+                    error("ERROR: Expired token.", context.Response);
+                    return;
+                }
+
+                Console.WriteLine(
+                    "[Authorization] Access for {0} to \"{1}\".",
+                    payload.fullName, context.Request.Url);
+
+                // Direct the module to handle the request
+                try {
+                    await module.HandleRequestAsync(context);
+                } catch (Exception e) {
+                    moduleError = e;
+                }
+
+            } catch (IntegrityException) {
+                error("ERROR: Invalid JWT signature.", context.Response);
+            } catch (Exception) {
+                error("Error: Failure to parse JWT.", context.Response);
+            }
+
+            if (moduleError != null) {
+                throw moduleError;
+            }
+        }
+
+        /// <summary>
+        /// Check whether the incoming Authorisation header is not null and contains the expected bearer token
+        /// </summary>
+        /// <param name="request"></param>
+        /// <param name="response"></param>
+        /// <returns></returns>
+        public static bool ValidateToken(IHttpRequest request, IHttpResponse response) {
+            var authorization = request.Headers["Authorization"];
+
+            if (string.IsNullOrEmpty(authorization)) {
+                error("ERROR: No authorization token provided.", response);
+                return false;
+            }
+
+            // I completely do not understand this pattern, and I want someone to tell me why. - Taylor
+            if (!authorization.StartsWith(bearerPrefix)) {
+                error("ERROR: Invalid authorization token provided.", response);
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Check if the user of the current session is signed in and if they contain the appropriate permissions
+        /// </summary>
+        /// <param name="permissions"></param>
+        /// <param name="request"></param>
+        /// <param name="response"></param>
+        /// <param name="jwtSecret"></param>
+        /// <returns></returns>
+        public static bool ValidatePermissions(List<string> permissions, IHttpRequest request, IHttpResponse response, string jwtSecret) {
```

Comment:
`ValidatePermissions` and `ValidateToken` have a lot of duplicate code from `OnRequestAsync`. Is there any way you can address this? Maybe have OnRequestAsync use these functions?

I recommend calling them `validatePermissions` and `validateToken` if possible 😓  I want to keep to the style where possible, and without a linter to point it out we have a problem.